elastic/kibana
# [Response Ops][Alerting] Skip settings and mapping updates on partial alerts indices.

Resolves https://github.com/elastic/kibana/issues/162630

## Summary

Adds checks for invalid alert index names in two places:
- before updating the underlying settings/mappings of a concrete index; attempting these updates would typically throw an error and prevent alert resource installation from successfully completing, causing subsequent writes to fail; in this PR ,we check for unexpected index names and log a warning
- before updating an alert document; there is a scenario where an existing active alert document could be in a partial or restored alert index and trying to update that alert document would fail. to prevent these failures, we skip the update and log a warning. we expect this case to be more rare as most times frozen indices only contain old alerts so the likelihood of it containing a currently active alert should be low.

## To Verify
- Run ES with these options: `yarn es snapshot --license trial --ssl -E path.data=../data_partial_alerts -E path.repo=<snaphot folder>-E xpack.searchable.snapshot.shared_cache.size=100b -E indices.lifecycle.poll_interval=1m`
- Start Kibana
- Create a snapshot repository here: https://localhost:5601/app/management/data/snapshot_restore/add_repository. Use `Shared File System` and use the same path as you used for `path.repo` when running ES
- Modify the `.alerts-ilm-policy` to roll over in the hot phase with max age of 3 minutes. Add a frozen phase that moves data into the frozen phase after 5 minutes.
- Create some rules that generate alerts. I did both metric threshold (uses lifecycle executor) and index threshold (uses framework).
- Wait for ILM to run and move indices to frozen. This will take a few ILM cycles but eventually you should be able to do a `GET .internal.alerts-stack.alerts-default-*/_alias/.alerts-stack.alerts-default` and see a partial index name in the results
- Restart Kibana. You should see warnings logged related to the partial indices but Kibana should successfully start and rule execution should succeed.

## Notes

I tested what would happen if we added a bunch of new fields to a component template and increased the total fields limit in the presence of partial indices. Here, it works in our favor that we only allow additive changes to our mappings, so the existing partial indices keep the old mappings and don't need a field limit update because their mappings don't change. Searching against both the alerts alias (that targets partial and normal indices) works as expected and searching directly against the partial index works as expected.
